instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/internal/ProtocolSpecificEvents.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.netty.v4_1.internal;
+
+import static io.opentelemetry.api.common.AttributeKey.stringArrayKey;
+import static io.opentelemetry.api.common.AttributeKey.stringKey;
+
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Adds events to {@link Span}s for the enumerated protocols and situations
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public enum ProtocolSpecificEvents {
+  /**
+   * The event after which point the server or client transmits or receives, respectively, in one of
+   * the signified upgraded protocols, per <a
+   * href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism">protocol
+   * upgrade mechanism</a>.
+   */
+  SWITCHING_PROTOCOLS("http.response.status_code.101.upgrade") {
+
+    @Override
+    public void addEvent(Context context, HttpRequest request, HttpResponse response) {
+      Span.fromContext(context)
```

Comment:
It's not described in the spec. I couldn't find anything relating to the handling of `101` and its somewhat unique nature. This event I added just to make sure it was captured _somewhere_ so anyone could observe that it happened and also test against it sanely. It could easily be done as an attribute, similar to `http.request.method_original` ([here](https://opentelemetry.io/docs/specs/semconv/attributes-registry/http/)) to capture whatever semantics of the original response. But, as it's the only status TMK that would require such handling, special-casing it seemed like wasted effort in the moment. And since the time this was sent is technically relevant, I opted for an event.

Reply:
our convention is to by default only use attributes that are in the spec, for consistency I think we should apply this also to events. Instrumentations use a flag such as `otel.instrumentation.grpc.experimental-span-attributes` to enable these experimental attributes, I guess we could use a similar flag here. Library instrumentations have a setter in the telemetry builder class to turn such features on. @trask wdyt? Has something like this been considered for the spec.

Reply:
From my end, as primarily a user, that makes sense, but in trying to implement a first pass at this I'm caught up on how to obtain that same value as grpc does: grpc uses `InstrumentationConfig` which is part of `javaagent-extension-api`, but the code paths in netty we'd need to update to use this are all in `library` modules. My understanding -- and please correct me if I'm wrong -- is that javaagent-extension-api module wouldn't be available or loaded when using library modules, necessarily. Is there an existing pattern for obtaining this similar `experimental-events` config value in some other way when shared between both instrumented and library code paths?

Reply:
Your understanding is correct `javaagent-extension-api` shouldn't be used in `library` modules. The usual pattern is that the `library` modules are configured programmatically. You could have a setter like https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/bc034f3b1b8d66e6c9f5a133407922278bc87086/instrumentation/grpc-1.6/library/src/main/java/io/opentelemetry/instrumentation/grpc/v1_6/GrpcTelemetryBuilder.java#L121-L131 and propagate this boolean to the constructor of `HttpClientResponseTracingHandler`. In the `javaagent` module you'd use `InstrumentationConfig` to read the configuration property in `NettyClientSingletons`, add a getter for that value and pass it to the constructor of `HttpClientResponseTracingHandler`.

Reply:
> It's not described in the spec. I couldn't find anything relating to the handling of 101 and its somewhat unique nature.

this sounds like a good question to raise in https://github.com/open-telemetry/semantic-conventions

and in this repo we can hide any non-specified behavior behind an opt-in flag as @laurit mentions

Reply:
Sounds good -- I'll open that question when I get the chance over the next few days.

The implementation was a little less straightforward than grpc as the `*TelemetryBuilder`s weren't being universally used within Netty, but I made an attempt to bring them in line roughly with the grpc pattern. Hopefully that's on the right track!